File: pipeline/routers/escalations.py
"""routers/escalations.py - code/data-restriction escalations: list/stats/seen/override."""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from auth_dependency import require_auth
from escalation_store import (
    get_escalations,
    get_escalation,
    get_escalation_stats,
    mark_escalations_seen,
    mark_escalation_overridden,
)
from prompt_proposal_store import create_proposal_from_override

router = APIRouter()


@router.get("/escalations", dependencies=[Depends(require_auth)])
def list_escalations(status: Optional[str] = None, page: int = 1, limit: int = 200):
    escalations, total = get_escalations(status, page=page, limit=limit)
    return {"escalations": escalations, "total": total, "page": page, "limit": limit}


@router.get("/escalations/stats", dependencies=[Depends(require_auth)])
def escalations_stats():
    """Counts by type, computed by MongoDB - what EscalationSection's stat bar needs."""
    return get_escalation_stats()


class EscalationSeenPayload(BaseModel):
    ids: List[str]


@router.post("/escalations/seen", dependencies=[Depends(require_auth)])
def escalations_seen(payload: EscalationSeenPayload):
    count = mark_escalations_seen(payload.ids)
    return {"status": "ok", "updated": count}


class EscalationOverridePayload(BaseModel):
    note: str
    author: str


@router.post("/escalations/{escalation_id}/override", dependencies=[Depends(require_auth)])
def override_escalation(escalation_id: str, payload: EscalationOverridePayload):
    """
    A human disagrees with a triage verdict (most often "none") and wants
    a real prompt fix drafted from it. Drafts one against the CURRENT live
    prompt and creates a normal pending proposal from it - this endpoint
    never touches the live prompt itself, the resulting proposal still
    needs a separate Approve.
    """
    escalation = get_escalation(escalation_id)
    if escalation is None:
        raise HTTPException(status_code=404, detail="Escalation not found")

    proposal_id = create_proposal_from_override(escalation, payload.note, payload.author)
    if proposal_id is None:
        raise HTTPException(status_code=502, detail="Could not draft a fix from this override - try a more specific note.")

    mark_escalation_overridden(escalation_id, proposal_id)
    return {"status": "proposal_created", "proposal_id": proposal_id}

File: client/src/api/escalations.js
/**
 * api/escalations.js
 * ------------------
 * Code/data-restriction escalations from the triage agent - the
 * "Everything Else Reviewed" section. See server/routes/escalations.js.
 */

import { apiFetch, cachedGet, invalidateCache, invalidateTriageViews } from "./client.js";

/**
 * One page of code/data-restriction escalations from the triage agent.
 * Pass status ("unseen") and/or type ("none"/"code_restriction"/
 * "data_restriction") to filter, or triggerType ("draft_edit") for the
 * "Permanent Edit" filter. Returns {escalations, total, page, limit}.
 */
export async function fetchEscalations(status, { page = 1, limit = 200, type, triggerType } = {}) {
  const params = new URLSearchParams({ page: String(page), limit: String(limit) });
  if (status) params.set("status", status);
  if (type) params.set("type", type);
  if (triggerType) params.set("trigger_type", triggerType);
  return cachedGet(`/api/escalations?${params.toString()}`);
}

/** Counts by type for EscalationSection's stat bar, computed by MongoDB. */
export async function fetchEscalationStats() {
  return cachedGet("/api/escalations/stats");
}

/** Marks a batch of escalations as seen - called once the Escalations section is actually viewed. */
export async function markEscalationsSeen(ids) {
  const result = await apiFetch("/api/escalations/seen", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ids }),
  });
  invalidateCache("/api/escalations");
  invalidateCache("/api/notifications");
  return result;
}

/** Disagrees with a triage verdict - drafts a real prompt-fix proposal from a human note, added to the same pending queue (still needs its own Approve). */
export async function overrideEscalation(id, note, author) {
  const result = await apiFetch(`/api/escalations/${id}/override`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ note, author }),
  });
  invalidateTriageViews();
  return result;
}
